[sync BN]

jjsjann123 opened the pull request:
- Summary:

Added synchronized batch normalization, allows synchronization of stats across mini-batches between processes within a process group.
Current implementation uses a mixture of extended ATen native functions (cpp cuda extension) + torch.nn.modules (c10d python API)

- User-facing api:

1. torch.nn.utils.convert_sync_batchnorm(modules, process_group=None)

2. torch.nn.SyncBatchNorm(num_features, eps=1e-5, momentum=0.1, affine=True, track_running_stats=True, ***process_group=None***)
                           

- supported use case:
DistributedDataParallel with ***single-gpu multi-process***

a. User creates model containing `torch.nn.SyncBatchNorm` layers through one of the ways listed below:

  1. use layers directly:

     torch.nn.SyncBatchNorm(...)

     similar API as with torch.nn.BatchNormXd(...)
     with added argument `process_group` which is used to limit the scope of                                            
     synchronization within each process group. Default value is None, which                                            
     implies synchronization across all GPUs                                                                            

  2. use torch.nn.utils.convert_sync_batchnorm(modules, process_group)                                                  

     recursively convert all `torch.nn.BatchNormXd` into `torch.nn.SyncBatchNorm`                                       
     preserving values of parameters/buffers.                                                                           
     the utility function also allows user to specify process_group value to all                                        
     converted layers.                                                                                                  

b. user wraps their model with                                                                                          
   `torch.distributed.parallel.DataParallelDistributed`, from this point, user                                          
   should follow the general guidelines for DDP use guide                                                               

- Error checking

For use cases not supported, we error out:                                                                              

1. Application launched without ddp:                                                                                    
   > import torch                                                                                                     
   > sbn = torch.nn.SyncBatchNorm(10).cuda()                                                                          
   > inp = torch.randn(5, 10, 3, 3).cuda()                                                                            
   > sbn(inp) --> Error!                                                                                              
   > AttributeError: SyncBatchNorm is only supported within torch.nn.parallel.DistributedDataParallel                     

2. Application launched using DDP with multi-GPU per-process:                                                           
   > ddp_module = nn.parallel.DistributedDataParallel(module, device_ids=device_ids, output_device=args.local_rank)   
   > ValueError: SyncBatchNorm is only supported for DDP with single GPU per process

jjsjann123 committed:
[sync BN]

Summary:
Added synchronized batch normalization, allows synchronization of stats
across mini-batches between processes within a process group.

Current implementation uses a mixture of extended ATen native functions
(cpp cuda extension) + torch.nn.modules (c10d python API)

This is a WIP:
1. only supports GPU.
2. only supports single GPU per process.

jjsjann123 commented:
This is the first phase of this PR. We want to have sync BN support there first.
We can discuss support for multiple GPU per process as well as moving communication into pure c++ as well.
One thing we need to discuss here in this PR is testing.

My monkey python tests runs multiple processes & communications, I used it for functional.
For upstream tests, I'm thinking about utilizing unit tests for batch norm layer and adding some simple test in test_distributed.py to validate the communication/parallel welford part.

As I cannot find an official module upstream doing similar things, feedback or hints would be greatly appreciated.

jjsjann123 commented:
Pinging @SsnL for visibility.

weiyangfb commented:
can I ask at a very high level what strategy is using here to implement sync BN?

jjsjann123 commented:
@weiyangfb If i understand the question correctly, the implementation here is to: 1. calculate stats  (mean/var) for local mini-batch; 2. all reduce stats across all processes to calculate global mean/var; 3. Apply element-wise normalization.

Backwards follows identical logic with slightly different arithmetic.

jjsjann123 committed:
[SyncBatchNorm]
fixing backwards path with process group
fixing linter issue

jjsjann123 committed:
Merge remote-tracking branch 'origin/master' into sbn_PR

jjsjann123 commented:
Many failed tests. Seems like I got a lemon commit in master. Will merge again later.

ssnl reviewed (commented):
Thank you! Structure looks good to me. Kernels seem good too, although I didn't try to understand the details (e.g., math). I have some general questions, and this definitely needs some tests.

fmassa commented:
Quick question: doesn't this PR have some overlap with the work from @apaszke from https://github.com/pytorch/pytorch/pull/15146 ? Both exposes a new `batch_norm_update_stats` function.

jjsjann123 commented on torch/nn/_functions/thnn/normalization.py:
```diff
@@ -135,4 +135,101 @@ def backward(self, grad_output):
+
+        # calcualte mean/invstd for input.
+        mean, invstd = torch.batch_norm_stats(input, eps)
+        if torch.distributed.is_initialized():
```
If the user used SyncBatchNorm in a single process environment, shouldn't we still want to support that instead of throwing an error?

SsnL replied:
Hmm that would be good to have indeed. Could you instead check for `is_initialized & world_size > 1` and directly use `batch_norm` if the check fails (instead of using the split kernel).

jjsjann123 commented on torch/nn/_functions/thnn/normalization.py:
```diff
@@ -135,4 +135,101 @@ def backward(self, grad_output):
+            mean_l = [mean_all.narrow(0, i, 1) for i in range(world_size)]
+            invstd_l = [invstd_all.narrow(0, i, 1) for i in range(world_size)]
+            # using all_gather instead of all reduce so we can calculate mean/var in one go
+            torch.distributed.all_gather(mean_l, mean, process_group)
```
less nccl calls so technically yes. Realistically speaking the actual communication part is tiny in the overall timeline. The synchronization might have some impact here, but since we have two consecutive all_gather/all_reduce, combining the two calls doesn't really help.

jjsjann123 commented on torch/nn/_functions/thnn/normalization.py:
```diff
@@ -135,4 +135,101 @@ def backward(self, grad_output):
         return grad_input
 
 
+class SyncBatchNorm(Function):
```
Sure.
Should it go to `torch/nn/modules`?

SsnL replied:
Yes. Please put in `nn/modules`.

jjsjann123 commented on torch/nn/_functions/thnn/normalization.py:
```diff
@@ -135,4 +135,101 @@ def backward(self, grad_output):
+                    mean_dy, torch.distributed.ReduceOp.SUM, process_group)
+                mean_dy = mean_dy / world_size
+                torch.distributed.all_reduce(
+                    mean_dy_xmu, torch.distributed.ReduceOp.SUM, process_group)
```
Yeah, the kernels could use some fine tuning.
I'll put a TODO here.

jjsjann123 commented on torch/nn/_functions/thnn/normalization.py:
```diff
@@ -135,4 +135,101 @@ def backward(self, grad_output):
+                world_size = torch.distributed.get_world_size()
+            mean_all = torch.empty(world_size, mean.size(0), dtype=mean.dtype, device=mean.device)
+            invstd_all = torch.empty(world_size, invstd.size(0), dtype=invstd.dtype, device=invstd.device)
+            mean_l = [mean_all.narrow(0, i, 1) for i in range(world_size)]
```
Will switch.
Looking at the API, `unbind` is a simplified `narrow`. There shouldn't be much more overhead using `narrow` comparing to using `unbind`?

SsnL replied:
There shouldn't be much more overhead, but it will look nicer :)

jjsjann123 commented:
I use Welford to calculate mean/var in a single pass as well. Welford has better numerical characteristics, which is desired.

The all_gather that is called here is only going to gather 1 set of intermediate mean/m2n per process, for a reasonable cluster size there shouldn't be much difference between all_gather/all_reduce

jjsjann123 committed:
Merge remote-tracking branch 'origin/master' into sbn_PR

jjsjann123 committed:
Merge remote-tracking branch 'origin/master' into sbn_PR

jjsjann123 committed:
addressing review comments:

1. fallback to batch_norm when sync is not required;
2. inplace operator to save memory;
3. swtich from narrow to use unbind.

jjsjann123 committed:
moving SyncBatchNorm function from thnn to modules

jjsjann123 committed:
removing trailing whitespace

jjsjann123 committed:
test added for sync batch norm layer

jjsjann123 committed:
fixing pylint issue with new line

apaszke commented on torch/nn/modules/_functions.py:
```diff
@@ -0,0 +1,90 @@
+        if self.needs_input_grad[0]:
+            # synchronizing stats used to calculate input gradient.
+            # TODO: move div_ into batch_norm_backward_elemt kernel
+            if torch.distributed.is_initialized():
```
That's a pretty flaky check for being used within DDP. It's not the only case when people use distributed, and it might cause deadlocks for people.

Also, we should consider doing those asynchronously if possible, so we could unblock other branches of the backwards pass.

jjsjann123 replied:
I should have removed `if torch.distributed.is_initialized():` here, since I was doing the check in `SyncBatchNorm` module already.

So the follow up question is, what should be the proper way to check (here https://github.com/pytorch/pytorch/pull/14267/files#diff-98c9603e09a23834975e9d62c07c30d9R427) whether distributed is used?

I didn't quite get 
>  we should consider doing those asynchronously if possible
Are you referring to the all_reduce/gather here that should be done asynchronously?

apaszke replied:
> Are you referring to the all_reduce/gather here that should be done asynchronously?

Yes, the new distributed module can return futures, but after giving it some though it's probably not doable since you have to have valid CUDA memory for next kernels at this point.

> So the follow up question is, what should be the proper way to check (...) whether distributed is used?

From what I understand this module is supposed to be used from within `DistributeDataParallel` only. You could define a thread-local Python variable that is set when `DistributedDataParallel.forward` executes, and is unset otherwise. Then, check it inside this function.

apaszke commented on torch/nn/modules/batchnorm.py:
```diff
@@ -316,3 +317,125 @@ def _check_input_dim(self, input):
+                process_group = self.process_group
+            world_size = torch.distributed.get_world_size(process_group)
+
+        # fallback to framework BN when synchronization is not necessary
```
Hmmm I'd also expect it to work from within `nn.DataParallel`, but we don't even check if it's used like that.

jjsjann123 replied:
nn.DataParallel is not supported in this PR yet.
DataParallel would require communication among GPUs within a single process. I'm not aware of proper upstream support for that. So we decide to upstream this single-gpu DDP sync bn first.

apaszke replied:
I think all distributed functions you have here have been present in multi-GPU versions since 0.1 in `torch.cuda.comm`. My issue with the current implementation is that it's extremely non-composable which is why I'm skeptical about including it in the core library. Such silent assumptions about where a module can or cannot be used make it very hard for one to reason about their code.

apaszke replied:
Basically, if you're going about adding a non-composable block, you should at least make it work only in the cases you implemented, and everything else should be a hard error. This is not the case in here.

jjsjann123 replied:
Agreed that I should have throw a loud error in cases not supported.
I'll add a check and error out when multi-GPU per process is detected when executing SyncBatchNorm

jjsjann123 committed:
Merge remote-tracking branch 'origin/master' into sbn_PR

jjsjann123 committed:
update on conflicted C++ api

jjsjann123 committed:
Merge remote-tracking branch 'origin/master' into sbn_PR

jjsjann123 committed:
DistributedDataParallel condition check for SyncBatchNorm

1. added check so that SyncBatchNorm is only supported for single GPU per
process run with DistributedDataParallel.
2. added utility function to convert BatchNorm layer in module to SyncBatchNorm
layer

jjsjann123 commented:
@apaszke Added check during DDP initialization so that SyncBatchNorm is only supported with single GPU per process with DDP launch.

jjsjann123 committed:
pylint fix

jjsjann123 committed:
removing obsolete distributed initialization check

jjsjann123 commented:
Now flake8 is complaining about something that is totally made up.

./test/common_utils.py:748:1: E302 expected 2 blank lines, found 1
./test/test_torch.py:1235:5: E303 too many blank lines (2)

Blank lines seem to be fine in the repo, and my commits didn't touch them at all.

jjsjann123 commented:
@pytorchbot retest this please

jjsjann123 committed:
Merge remote-tracking branch 'origin/master' into sbn_PR

jjsjann123 commented:
@soumith Summary updated at the top in the PR message

soumith commented:
thanks, will take a look.

soumith commented:
cc: @mrshenli @teng-li and @pietern for review.
@jjsjann123 can you also check and update what happens if this is accidentally used by user in a nn.DataParallel setting

jjsjann123 commented:
Launching with `nn.DataParallel` would be the case where it's not launched through DDP, at `SyncBatchNorm.forward()` an exception is raise:

1. Application launched without ddp:
> import torch
> sbn = torch.nn.SyncBatchNorm(10).cuda()
> inp = torch.randn(5, 10, 3, 3).cuda()
> sbn(inp) --> Error!
> AttributeError: SyncBatchNorm is only supported within torch.nn.parallel.DistributedDataParallel

pietern commented on test/test_distributed.py:
```diff
@@ -1357,6 +1374,86 @@ def test_DistributedDataParallel(self):
+        for name, child in module.named_children():
+            mod.add_module(name, self._convert_syncbn_model(child))
+        del module
+        return mod
```
The mod vs module naming is confusing. Input/output or something else like it is more appropriate IMO.

jjsjann123 replied:
will rename. Thanks.

pietern commented on torch/nn/modules/_functions.py:
```diff
@@ -0,0 +1,89 @@
+        invstd_l = list(invstd_all.unbind(0))
+        # using all_gather instead of all reduce so we can calculate mean/var in one go
+        torch.distributed.all_gather(mean_l, mean, process_group)
+        torch.distributed.all_gather(invstd_l, invstd, process_group)
```
You can pass async_op=True to run both in parallel instead of serially to improve latency.

Make sure to capture the returned object and call `wait()` on it, before using the output tensors though.

jjsjann123 replied:
Sounds good. Will do.

pietern commented on torch/nn/modules/_functions.py:
```diff
@@ -0,0 +1,89 @@
+            mean_dy.div_(world_size)
+            torch.distributed.all_reduce(
+                mean_dy_xmu, torch.distributed.ReduceOp.SUM, process_group)
+            mean_dy_xmu.div_(world_size)
```
These allreduce calls can interfere with ones kicked off by DDP itself.

If autograd runs with single threaded with deterministic ordering you'll be fine, but as soon as it doesn't (e.g. there are multiple branches of the forward graph where the backward functions can be called in parallel), you'll run into deadlocks. This can be avoided by creating a new process group from the main one with `new_group` and using that throughout. Note that having multiple of these sync batch norm layers running `backward` in parallel can still deadlock, or worse, result in mixed up data, so for guaranteed correctness you'll have to use a separate process group per sync batch norm layer. This is not ideal and we may need to find a different solution for this.

cc @mrshenli @teng-li

jjsjann123 replied:
This is a gotcha for me. Thanks a lot for pointing out the issue. I don't fully understand how the allreduce call would cause trouble in branching while DDP handles that fine. Maybe I'll ask more details about this in private channel.

Just to reiterate, for the time being, it's a safe WAR as long as I have a separate process group per sync batch norm layer and use it for *both* forward and backward pass.
I'll copy/create a process group inside the initializer of SyncBatchNorm.

jjsjann123 replied:
Realized that duplicating process_groups in the initializer of SyncBatchNorm would not work.

As new_group should be called by all processes in the main group, and inside the initializer or converter function, each process would only see the given process_group it belongs to for that layer.

alekhka replied:
I have this exact problem now!
My model has many SyncBNs in parallel (I also have an atypical module where I call autograd.grad() repeatedly on two tensors whose graph has syncBNs). In short the `backward()` has many `all_reduce` calls and some run in parallel. With DDP the code hangs without error and with NCCL logs I can see that it is because `all_reduce` hangs. 

I have tried torch native SyncBN, apex syncBN, and @ppwwyyxx 's NaiveSyncBN. This happens in both torch DDP and apex DDP (even with `delay_allreduce=True`). 

Is there something I can do to fix it? @jjsjann123 @pietern

pietern commented on torch/nn/modules/batchnorm.py:
```diff
@@ -316,3 +317,147 @@ def _check_input_dim(self, input):
+                             .format(input.dim()))
+
+    def _specify_process_group(self, process_group):
+        self.process_group = process_group
```
This function is not used.

pietern commented on torch/nn/utils/sync_batch_norm.py:
```diff
@@ -0,0 +1,45 @@
+    for name, child in module.named_children():
+        mod.add_module(name, convert_sync_batchnorm(child))
+    del module
+    return mod
```
Same mod vs module comment as earlier one.

jjsjann123 replied:
Noted

pietern reviewed (commented):
There is a potential for deadlock with the current calls to allreduce, as they execute in parallel with autograd, and therefore the primary DDP reduction hooks.

jjsjann123 committed:
Merge remote-tracking branch 'origin/master' into sbn_PR

jjsjann123 committed:
addressing review comments:
1. adding async_op for all_reduce calls
2. renaming variables
3. removing redundant code

jjsjann123 committed:
resolving pylint issue

ezyang commented:
We've decided we're going to go ahead and land this, and keep an eye on it for any problems that may occur later.

ezyang commented:
@pytorchbot rebase this please

Commit:
Merge remote-tracking branch 'origin/master' into HEAD

jjsjann123 commented:
Failures look scary, should I be concerned?

ezyang commented:
Ah, the PR bitrotted. Just a moment please.

ezyang committed:
s/int64_t/int/

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Undo incorrect change

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

LinxiFan commented:
How does torch 1.1's SyncBN feature compare to Nvidia [apex](https://github.com/NVIDIA/apex) library?

soumith commented:
it's similar in functionality, but likely better perf because it directly integrates with nn.BatchNorm
